rpc_jobs/build_gating_venv.yml

```diff
@@ -17,108 +17,120 @@
         // can't use common.shared_slave as it calls install_ansible,
         // which calls this job.
         // can't use globalWraps because it calls common.shared_slave
-        lock('build_gating_venv'){
-          node('pubcloud_multiuse'){
-            try{
+        node('pubcloud_multiuse'){
+          try{
+            withCredentials(artifact_build.get_rpc_repo_creds()) {
+
               deleteDir()
               common.clone_with_pr_refs('rpc-gating',
                                         'git@github.com:rcbops/rpc-gating',
                                         env.RPC_GATING_BRANCH)
               dir('rpc-gating'){
                 // run in docker to ensure all the required apt packages are available
-                container = docker.build env.BUILD_TAG.toLowerCase()
+                container = docker.build(env.BUILD_TAG.toLowerCase())
               }
-              // Venv is build within docker as it requires extra packages
+
+              // Venv is built within docker, includes all the environment variables
               container.inside{
-                sh """#!/bin/bash -xeu
-
-                # Install virtualenv if required
-                if [[ ! -d ".venv" ]]; then
-                  requirements="virtualenv==15.1.0"
-                  pip install -U "\${requirements}" \
-                    || pip install --isolated -U "\${requirements}"
-
-                  if which scl
-                  then
-                    # redhat/centos
-                    source /opt/rh/python27/enable
-                    virtualenv --no-pip --no-setuptools --no-wheel --python=/opt/rh/python27/root/usr/bin/python .venv
-                    # hack the selinux module into the venv
-                    cp -r /usr/lib64/python2.6/site-packages/selinux .venv/lib64/python2.7/site-packages/ ||:
-                  else
-                    virtualenv --no-pip --no-setuptools --no-wheel .venv
+                sshagent(credentials:['rpc_repo_ssh_user_private_key']){
+                  sh """#!/bin/bash -xeu
+
+                  # Determine the SHA and archive name that we'll use
+                  pushd rpc-gating; SHA=\$(git rev-parse HEAD); popd
+                  archive="rpcgatingvenv_\${SHA}.tbz"
+
+                  # Set some SSH options
+                  SSH_OPTS="-o StrictHostKeyChecking=no"
+
+                  # Check to see if the virtual env archive already exists...if so, no need to rebuild it
+                  ssh \$SSH_OPTS \$REPO_USER@\$REPO_HOST "test -e \$REPO_PATH/\$archive"
+                  if [[ $? -eq 0 ]]; then
```

Comment:
`$?` needs escaping as otherwise groovy will interpret that `$`. This is causing the unit test fail.